pkg/util/sketch/topk_test.go

```diff
@@ -0,0 +1,369 @@
+package sketch
+
+import (
+	"bufio"
+	"container/heap"
+	"github.com/alicebob/miniredis/v2/hyperloglog"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"io"
+	"math/rand"
+	"net/http"
+	"sort"
+	"strconv"
+	"testing"
+	"time"
+)
+
+type event struct {
+	name  string
+	count int
+}
+
+func TestTopkCardinality(t *testing.T) {
+	max := 1000000
+	topk, err := newCMSTopK(100, 10, 10)
+	assert.NoError(t, err)
+	for i := 0; i < max; i++ {
+		topk.Observe(strconv.Itoa(i))
+	}
+	c, bigEnough := topk.Cardinality()
+	// hll has a typical error accuracy of 2%
+	assert.True(t, (c >= uint64(float64(max)*0.98)) && (c <= uint64(float64(max)*1.02)))
+	assert.False(t, bigEnough)
+
+	topk, err = NewCMSTopkForCardinality(nil, 100, max)
+	assert.NoError(t, err)
+	for i := 0; i < max; i++ {
+		topk.Observe(strconv.Itoa(i))
+	}
+	c, bigEnough = topk.Cardinality()
+	assert.True(t, bigEnough)
+}
+
+// TODO: merging is not as accurate as it should be
+func TestTopK_Merge(t *testing.T) {
+	nStreams := 1000
+	k := 100
+	maxPerStream := 1000
+	events := make([]event, 0)
+	max := int64(0)
+
+	for i := 0; i < nStreams-k; i++ {
+		num := int64(maxPerStream)
+		n := rand.Int63n(num) + 1
```

Comment:
This test is not repeatable and it would be a pain to debug. Please use a rand instance with a constant seed.